_posts/2019/2019-08-20-end-to-end-tests-with-geb-and-spock.md

```diff
@@ -0,0 +1,154 @@
+still an issue with Geb, in that it does not produce screenshots properly in some cases, if `@Retry` is being used. But this is
+actually not a big problem, because nowadays we use video recording to understand failures, rather than screenshots.
+
+Be aware of the difference between repeating a test and retrying a test. When running stability tests, we do not want to retry any
+tests, because we do want to see the errors in this phase, so that we have a chance to fix them.
+
+
+## Concurrent test execution
+
+We had to take some precautions in order to be able to run tests concurrently.
+The word "concurrent" has two different meanings here. On one hand it refers to starting the program in a way that makes it invoke
+several tests in parallel. This can be achieved using the [`maxParallelForks`](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html#org.gradle.api.tasks.testing.Test:maxParallelForks) property of a Gradle test task.
+This way of parallelising tests written in JVM / Gradle does not require any specific support by the test framework.
+
+The other meaning of "concurrent" refers to having two instances of the same test running *somewhere*.
+For example, a test is running in some CI pipeline, while a developer is starting the same test on his own computer.
+Why would this be a problem?
+Consider the case that during the test, some credentials are being used to log in to some third-party website or service. This could
+have the effect of invalidating any simultaneous session that might be going on, or otherwise interfering with the test result.
+
+Often times our third-party integrations would allow us to create any number of test accounts, but there could be exceptions.
+In one case we were not allowed to do that, as we were testing against the live system of that integration, because no "sandbox" existed.
+
+So we had to come up with a mechanism, that would allow us to prevent concurrent execution of this test, and instead wait for the
+ongoing run to finish. Conveniently, Java already provides [an interface for this purpose](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html), which has exactly the desirable property that requesting a lock includes waiting for its availability, while
+disabling the current thread. Nice!  
+The interface is readily usable implemented by the [Redisson](https://github.com/redisson/redisson) library, which uses [Redis](https://redislabs.com/why-redis/) as its backend.
+
+
+## Diagnosing actual problems
+
+Of course, at some point, we want our end-to-end tests to be able to bring real problems with our software to our attention.
+Once random stability failures were mostly understood, this is working well for us.
+
+The tests helped us to notice several issues related to third-party integrations and the checkout and order processes.
```

Comment:
... as well as the checkout and order processes.

Reply:
Hier ist sowohl als auch gemeint. Drittpartien werden eingebunden, *während* der Checkout- und Order-Prozess durchlaufen wird.

Reply:
Jep :) That's exactly what it also says when using "as well as".